File: leetcode/js/number-of-good-pairs.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var numIdenticalPairs = function (nums) {
  let result = 0;

  for (let i = 0; i < nums.length - 1; i++) {
    result += nums.slice(i + 1).filter((num) => num === nums[i]).length;
  }

  return result;
};

var numIdenticalPairs = function (nums) {
  let result = 0;
  const countObj = {};

  for (const num of nums) {
    if (num in countObj) {
      result += countObj[num];
      countObj[num]++;
    } else {
      countObj[num] = 1;
    }
  }

  return result;
};

File: leetcode/rust/number-of-good-pairs.rs
impl Solution1 {
    pub fn num_identical_pairs(nums: Vec<i32>) -> i32 {
        let mut result = 0;

        for i in 0..(nums.len() - 1) {
            for j in (i + 1)..nums.len() {
                if nums[i] == nums[j] {
                    result += 1;
                }
            }
        }

        result
    }
}

impl Solution2 {
    pub fn num_identical_pairs(nums: Vec<i32>) -> i32 {
        use std::collections::HashMap;

        let mut map = HashMap::new();
        let mut count = 0;

        for num in nums {
            let val = map.entry(num).or_insert(0);
            count += *val;
            *val += 1
        }

        count
    }
}
